---
title: "ATACseq_Peaks_Analysis"
author: "Kfir Inbal"
date: "2024-07-27"
output: html_document
---
```{r}
#Load Libraries
library(DiffBind)
library(parallel)
library(tidyverse)
library(dplyr)
library(rtracklayer)
library(GenomicRanges)
library(ComplexHeatmap)
library(circlize)
library(RColorBrewer)
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)  # Replace with your genome of interest
library(org.Hs.eg.db)  # Replace with your organism of interest
library(clusterProfiler)
library(pheatmap)
library(fgsea)
library(enrichplot)
library(msigdbr)
library(ggrepel)

```

```{r}
#Functions:
pheatmap.scale <- function(x) {
  m = apply(x, 1, mean, na.rm = T)
  s = apply(x, 1, sd, na.rm = T)
  return((x - m) / s)
}

add_gene_names <- function(df_to_add_gene_names, dba_object) {
    # Step 1: Annotate Peaks with Gene Names
  txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene  # Replace with appropriate TxDb
  
  # Retrieve the peaks from the dba object
  peaks <- dba.peakset(dba_object, bRetrieve=TRUE)
  
  # Ensure that the sequence levels (chromosome names) match
  # Check chromosome names in your peaks
  peak_chr <- seqlevels(peaks)
  # Check chromosome names in your TxDb object
  txdb_chr <- seqlevels(txdb)
  
  # Print the chromosome names to understand the format
  #print(peak_chr)
  #print(txdb_chr)
  
  # Convert chromosome names in peaks if necessary
  # For example, if peaks have "1" and txdb has "chr1"
  if (!all(peak_chr %in% txdb_chr)) {
    seqlevelsStyle(peaks) <- seqlevelsStyle(txdb)  # Match styles
  }
  
  # Recheck the chromosome names
  #print(seqlevels(peaks))
  
  # Annotate the peaks
  peakAnno <- annotatePeak(peaks, tssRegion=c(-3000, 3000), TxDb=txdb, annoDb="org.Hs.eg.db")
  
  # Extract gene names
  geneNames <- peakAnno@anno$SYMBOL
  
  # Step 2: Add Gene Names to the dba Object
  dba_object$gene <- geneNames
  
  peakAnno_df <- as.data.frame(peakAnno)
  
  peakAnno_df_genes <- peakAnno_df[, c("SYMBOL","seqnames", "start", "end")]
  
  DF <- as.data.frame(df_to_add_gene_names)
  
  if (!all(DF$seqnames %in% txdb_chr)) {
    DF$seqnames <- paste0("chr", DF$seqnames)
  }
  
  # Merge the data matrix with gene annotation data
  DF_merged <- DF %>%
    left_join(peakAnno_df_genes, by = c("seqnames", "start", "end"))

  return(DF_merged)
}
```

```{r}


ATAC_Seq_DBS_Analysis <- function(Sample_Sheet_csv_file, celltype, group1, group2, DE_Method = DBA_EDGER, Normalization_Method = DBA_NORM_TMM, usePval = F, pval_th = 0.01, FC_th = -log(0.7), Capped_th = 5, outputdir = "./", dba_counts_RDS_File = "") {
  
  if (dba_counts_RDS_File == "") {
    # Load the sample sheet
    sampleSheet_data <- read.csv(Sample_Sheet_csv_file)
    
    # Initialize the dba object with the sample sheet
    dba <- dba(sampleSheet=sampleSheet_data) 
    dba_blklst <- dba.blacklist(dba) 
    dba_Counts <- dba.count(dba_blklst) #bParallel = TRUE, score=DBA_SCORE_READS, summits=75) 
    saveRDS(dba_Counts,file = paste0(outputdir, celltype, "_Counts.RDS"))
  } else {
    dba_Counts <- readRDS(dba_counts_RDS_File)
  }
  
  dba_Counts_norm <- dba.normalize(dba_Counts, method = DE_Method, normalize = Normalization_Method) 
  dba_Counts_norm_contrast <- dba.contrast(dba_Counts_norm, contrast = c("Condition", group1, group2))
  dba_Counts_norm_contrast_analyzed <- dba.analyze(dba_Counts_norm_contrast, method = DE_Method)
  
  png(paste0(outputdir,celltype,'_peak_counts_correlation_heatmap_',DE_Method,'_',Normalization_Method,'.png'), units="in", width=7, height=7, res=800)
  dba.plotHeatmap(dba_Counts_norm_contrast_analyzed, main = paste0("Correlation Plot for ",celltype ," Samples"))
  dev.off()
  
  pdf(paste0(outputdir,celltype,'_peak_counts_correlation_heatmap_',DE_Method,'_',Normalization_Method,'.pdf'), width=7, height=7)
  dba.plotHeatmap(dba_Counts_norm_contrast_analyzed, main = paste0("Correlation Plot for ",celltype ," Samples"))
  dev.off()
  
  png(paste0(outputdir,celltype,'_peak_counts_PCA_',DE_Method,'_',Normalization_Method,'.png'), units="in", width=7, height=7, res=800)
  dba.plotPCA(dba_Counts_norm_contrast_analyzed,DBA_CONDITION, label = DBA_ID)
  # Add a title using grid.text
  grid.text(paste0("PCA Plot for ",celltype ," Samples"), x = 0.5, y = 0.95, gp = gpar(fontsize = 16))
  dev.off()
  
  
  pdf(paste0(outputdir,celltype,'_peak_counts_PCA_',DE_Method,'_',Normalization_Method,'.pdf'),width=7, height=7)
  dba.plotPCA(dba_Counts_norm_contrast_analyzed,DBA_CONDITION, label = DBA_ID)
  # Add a title using grid.text
  grid.text(paste0("PCA Plot for ",celltype ," Samples"), x = 0.5, y = 0.95, gp = gpar(fontsize = 16))
  dev.off()
  
  #Reporting SIGNIFICANT differential Binding Sites
  DBS_sig <- dba.report(dba_Counts_norm_contrast_analyzed,th = pval_th, contrast = 1,bUsePval=usePval, method=DE_Method, bCounts=TRUE)
  
  write.csv(as.data.frame(DBS_sig),paste0(outputdir,celltype,'_Sig_Counts_norm_contrast_analyzed_DBS',DE_Method,'_',Normalization_Method,'.csv'))
  
  
  DBS_sig <- add_gene_names(DBS_sig, dba_Counts_norm_contrast_analyzed)
  
  write.csv(DBS_sig, paste0(outputdir,celltype,'_Genes_Sig_Counts_norm_contrast_analyzed_DBS',DE_Method,'_',Normalization_Method,'.csv'))
  
  #########       Volcano plot
  
  #All differential binding sites
  DBS_All <- as.data.frame(dba.report(dba_Counts_norm_contrast_analyzed, th = 1, contrast = 1, bUsePval=usePval, method=DE_Method, bCounts=TRUE))
  
  write.csv(DBS_All, paste0(outputdir,celltype,'_ALL_Counts_norm_contrast_analyzed_DBS',DE_Method,'_',Normalization_Method,'.csv'))
  
  DBS_All <- add_gene_names(DBS_All, dba_Counts_norm_contrast_analyzed)
  
  write.csv(DBS_All, paste0(outputdir,celltype,'_Genes_ALL_Counts_norm_contrast_analyzed_DBS',DE_Method,'_',Normalization_Method,'.csv'))
  
  # Define significance thresholds
  pval_threshold <- pval_th
  fold_change_threshold <- FC_th
  
  # Create a new column to classify points based on significance
  DBS_All$Significance <- with(DBS_All, ifelse(p.value < pval_threshold & abs(Fold) > fold_change_threshold, "Significant", "Not Significant"))
  
  # Calculate -log10 p-value
  DBS_All$negLog10Pval <- -log10(DBS_All$p.value) # or use P.value if you prefer p-values over FDR
  
  # Cap -log10(p-value) at 5 for plotting
  DBS_All$negLog10Pval_capped <- ifelse(DBS_All$negLog10Pval > Capped_th, Capped_th, DBS_All$negLog10Pval)
  # Create a new column to classify points that are capped
  DBS_All$Capped <- ifelse(DBS_All$negLog10Pval > Capped_th, "Capped", "Not Capped")
  
  # Identify top significant genes
  top_genes <- DBS_All[DBS_All$Significance == "Significant",]
  top_genes <- top_genes[order(top_genes$p.value), ] # Order by FDR
  top_genes <- head(top_genes, 10) # Select top 10 genes
  
  # Create the volcano plot using ggplot2
  volcano_plot <- ggplot(DBS_All, aes(x = Fold, y = negLog10Pval_capped, color = Significance, shape = Capped)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  scale_color_manual(values = c("Significant" = "red", "Not Significant" = "grey")) +
  scale_shape_manual(values = c("Capped" = 17, "Not Capped" = 16)) + # Different shapes for capped points
  geom_hline(yintercept = -log10(pval_threshold), col = "red", linetype = "dashed") +
  geom_vline(xintercept = c(-fold_change_threshold, fold_change_threshold), col = "blue", linetype = "dashed") +
  geom_text_repel(data = top_genes, aes(label = SYMBOL), size = 3) +
  labs(title = paste0("Volcano Plot Of ",celltype," - ", group1," vs ", group2, " DBS (Approximated DEG)"), x = "Log2 Fold Change", y = "-log10(Pvalue)", color = "Gene Significance", shape = "Capped Values") +
  ylim(0, 5) +
  annotate("text", x = max(DBS_All$Fold), y = 4.5, label = "Red: Significant genes\n(pvalue < 0.05 and abs(log2Fold) > log2(0.7))", hjust = 1, size = 3)
  
  png(paste0(outputdir,celltype,'_Volcano_Plot_',DE_Method,'_',Normalization_Method,'.png'), units="in", width=7, height=7, res=800)
  print(volcano_plot)
  dev.off()
  
  pdf(paste0(outputdir,celltype,'_Volcano_Plot_',DE_Method,'_',Normalization_Method,'.pdf'), width=7, height=7)
  print(volcano_plot)
  dev.off()
  
  results_list <- list(c(dba_counts = list(dba_Counts), dba_Counts_norm = list(dba_Counts_norm), dba_Counts_norm_contrast = list(dba_Counts_norm_contrast), dba_Counts_norm_contrast_analyzed = list(dba_Counts_norm_contrast_analyzed), DBS_sig_report = list(DBS_sig), DBS_All_report = list(DBS_All), VolcanoPlot = list(volcano_plot)))  
  saveRDS(object = results_list, file = paste0(outputdir,celltype,'_Results_List_',DE_Method,'_',Normalization_Method,'.RDS'))
 return(results_list)
  

} 


```


```{r}
#How to run:
#
#CD4_Good_Poor_Results_List <- ATAC_Seq_DBS_Analysis("Sample_Sheet_CD4.csv", "CD4", "good", "poor", usePval = T, pval_th = 0.05, outputdir = "./CD4_GoodvsPoor_Results")
#
#CD4_Good_Poor_Results_List <- ATAC_Seq_DBS_Analysis("Sample_Sheet_CD4.csv", "CD4", "good", "poor", usePval = T, pval_th = 0.05, outputdir = "./CD4_GoodvsPoor_Results", dba_counts_RDS_File = "./CD4_Counts.RDS")
#
#CD8_Good_Poor_Results_List <- ATAC_Seq_DBS_Analysis("Sample_Sheet_CD8.csv", "CD8", "good", "poor", usePval = T, pval_th = 0.05, outputdir = "./CD8_GoodvsPoor_Results")
#
#CD8_Good_Poor_Results_List <- ATAC_Seq_DBS_Analysis("Sample_Sheet_CD8.csv", "CD8", "good", "poor", usePval = T, pval_th = 0.05, outputdir = "./CD8_GoodvsPoor_Results/", dba_counts_RDS_File = "./CD8_GoodvsPoor_Results/CD8_GoodvsPoor_ResultsCD8_Counts.RDS")


```

```{r}
#CD4_Good_Poor_Results_List <- ATAC_Seq_DBS_Analysis("Sample_Sheet_CD4.csv", "CD4", "good", "poor", usePval = T, pval_th = 0.05, outputdir = "./CD4_GoodvsPoor_Results/")
CD4_Good_Poor_Results_List <- ATAC_Seq_DBS_Analysis("Sample_Sheet_CD4.csv", "CD4", "good", "poor", usePval = T, pval_th = 0.05, outputdir = "./CD4_GoodvsPoor_Results/", dba_counts_RDS_File = "./CD4_GoodvsPoor_Results/CD4_Counts.RDS")

#CD4_Good_Poor_Results_List_Deseq2 <- ATAC_Seq_DBS_Analysis("Sample_Sheet_CD4.csv", "CD4", "good", "poor", usePval = T, pval_th = 0.05,DE_Method=DBA_DESEQ2, Normalization_Method = DBA_NORM_RLE , outputdir = "./CD4_GoodvsPoor_Results_Deseq/", dba_counts_RDS_File = "./CD4_GoodvsPoor_Results/CD4_Counts.RDS")
```

```{r}


CD4_Counts_norm_contrast_analyzed <- CD4_Good_Poor_Results_List[[1]]$dba_Counts_norm_contrast_analyzed
CD4_DBS_sig <- CD4_Good_Poor_Results_List[[1]]$DBS_sig_report
CD4_DBS_All <- CD4_Good_Poor_Results_List[[1]]$DBS_All_report

pvals <- dba.plotBox(CD4_Counts_norm_contrast_analyzed,contrast = 1,bUsePval = T ,method = DBA_EDGER)
venn_Gain_Loss <- dba.plotVenn(CD4_Counts_norm_contrast_analyzed, contrast = 1, method = DBA_EDGER,bUsePval =TRUE, bDB=TRUE,bGain=TRUE, bLoss=TRUE, bAll=FALSE, th=0.05,label1 = "Good", label2 = "Poor")

#Extracting Human TF genes
#Downloaded from AnimalTFDB4
tf_data <- read.table("Homo_sapiens_TF_AnimalTFDB4.txt", header = TRUE, sep = "\t")
CD4_DBS_All_TFs <- subset(CD4_DBS_All, CD4_DBS_All$SYMBOL %in% tf_data$Symbol)
write.csv(CD4_DBS_All_TFs[order(CD4_DBS_All_TFs$p.value,decreasing = FALSE),], "./CD4_GoodvsPoor_Results/CD4_Experiment_TF_Genes_ALL_Counts_norm_contrast_analyzed_DBSedgeRGLM_TMM.csv")

```

```{r}
##################################### CD4 GSEA #########################################

CD4_DBS_All_Volcano_Sig <- subset(CD4_DBS_All, Significance == "Significant")
# Remove duplicates by filtering rows with greatest absolute Foldchange for each gene
CD4_DBS_All_Volcano_Sig_NoDups <- CD4_DBS_All_Volcano_Sig %>%
  group_by(SYMBOL) %>%
  slice(which.max(abs(Fold))) %>%
  ungroup()

df_ForGSEA <- subset(CD4_DBS_All, select = c(Fold,p.value,SYMBOL))


C7_Pathways <- gmtPathways("c7.all.v2023.2.Hs.symbols.gmt")

# Prepare the data
ranks <- df_ForGSEA$Fold
names(ranks) <- df_ForGSEA$SYMBOL
head(ranks)
gene_list <- sort(ranks, decreasing = TRUE)
barplot(gene_list)

set.seed(123)
# Perform GSEA
fgseaRes <- fgsea(pathways = C7_Pathways,
                       stats = gene_list,
                       minSize = 15,
                       maxSize = 1000, eps = 0)#nperm = 10000)

# Create a subset where the pathway includes the substring "CD4"
fgseaRes_CD4 <- subset(fgseaRes, grepl("CD4", pathway))
fgseaRes_CD4 <- fgseaRes_CD4[order(fgseaRes_CD4$NES,decreasing = TRUE),]
fgseaRes_CD4_csv <- fgseaRes_CD4
fgseaRes_CD4_csv$leadingEdge <- sapply(fgseaRes_CD4_csv$leadingEdge, paste, collapse = ", ")

# Optionally, save these subsets to CSV files
write.csv(fgseaRes_CD4_csv, file = "./CD4_GoodvsPoor_Results/GSEA/CD4_Experiment_fgseaRes_CD4_C7gmt_sig.csv", row.names = FALSE)


```


```{r}

#All top pathways up and down regulated
fgseaRes$Enrichment = ifelse(fgseaRes$NES > 0, "Up-regulated", "Down-regulated")

topPathwaysUp <- na.omit(fgseaRes[NES > 0][order(pval)][1:10,])
topPathwaysDown <- na.omit(fgseaRes[NES < 0][order(pval)][1:10,])
topPathways <- rbind(topPathwaysUp, topPathwaysDown)

# Plot top pathways
#library(ggplot2)
#(topPathways, aes(reorder(pathway, NES), NES)) +
#  geom_col(aes(fill = NES > 0)) +
#  coord_flip() +
#  labs(x = "Pathway", y = "Normalized Enrichment Score", title = "Top GSEA Pathways") +
#  theme_minimal()


colos = setNames(c("firebrick2", "dodgerblue2"),
           c("Up-regulated", "Down-regulated"))
header = paste0("Top ",nrow(topPathways)," Enriched Pathways In CD4 Experiment (Total pathways: Up=", nrow(topPathwaysUp),", Down=",  nrow(topPathwaysDown), ")")


g1 = ggplot(topPathways, aes(reorder(pathway, NES), NES)) +
geom_point( aes(fill = Enrichment, size = size), shape=21) +
scale_fill_manual(values = colos ) +
scale_size_continuous(range = c(2,10)) +
geom_hline(yintercept = 0) +
coord_flip() +
labs(x="Pathway", y="Normalized Enrichment Score",
 title=header)

png("./CD4_GoodvsPoor_Results/GSEA/CD4_Experiment_topPathWays_GSEA_Plot.png", units="in", width=15, height=15, res=800)
g1
dev.off()


pdf("./CD4_GoodvsPoor_Results/GSEA/CD4_Experiment_topPathWays_GSEA_Plot.pdf", width=15, height=15)
g1
dev.off()
```


```{r}
#CD4 related pathways
#Removing not important pathways:
fgseaRes_CD4_filt_NonImportant <- fgseaRes_CD4[-which(fgseaRes_CD4$pathway %in% fgseaRes_CD4[c(4,8,9,10,11,15,18,19),]$pathway),]

fgseaRes_CD4_top30 = fgseaRes_CD4_filt_NonImportant[1:30,]
fgseaRes_CD4_top30$Enrichment = ifelse(fgseaRes_CD4_top30$NES > 0, "Up-regulated", "Down-regulated")

fgseaRes_CD4_top30 = rbind(head(fgseaRes_CD4_top30, n = nrow(subset(fgseaRes_CD4_top30, NES>0))),
                tail(fgseaRes_CD4_top30, n = nrow(subset(fgseaRes_CD4_top30, NES<0)) ))
total_up = sum(fgseaRes_CD4_top30$Enrichment == "Up-regulated")
total_down = sum(fgseaRes_CD4_top30$Enrichment == "Down-regulated")
header = paste0("Top ",nrow(fgseaRes_CD4_top30)," Enriched CD4 related Pathways In CD4 Experiment (Total pathways: Up=", total_up,", Down=",  total_down, ")")

colos = setNames(c("firebrick2", "dodgerblue2"),
           c("Up-regulated", "Down-regulated"))

png("./CD4_GoodvsPoor_Results/GSEA/CD4_Experiment_CD4Pathways_top30_GSEA_Plot.png", units="in", width=15, height=15, res=800)
ggplot(fgseaRes_CD4_top30, aes(reorder(pathway, NES), NES)) +
geom_point( aes(fill = Enrichment, size = size), shape=21) +
scale_fill_manual(values = colos ) +
scale_size_continuous(range = c(2,10)) +
geom_hline(yintercept = 0) +
coord_flip() +
labs(x="Pathway", y="Normalized Enrichment Score",
 title=header)
dev.off()


pdf("./CD4_GoodvsPoor_Results/GSEA/CD4_Experiment_CD4Pathways_top30_GSEA_Plot.pdf", width=15, height=15)
ggplot(fgseaRes_CD4_top30, aes(reorder(pathway, NES), NES)) +
geom_point( aes(fill = Enrichment, size = size), shape=21) +
scale_fill_manual(values = colos ) +
scale_size_continuous(range = c(2,10)) +
geom_hline(yintercept = 0) +
coord_flip() +
labs(x="Pathway", y="Normalized Enrichment Score",
 title=header)
dev.off()



```










```{r}

head(fgseaRes[order(padj, -abs(NES)), ], n=1)
plotEnrichment(C7_Pathways[["GSE26488_HDAC7_KO_VS_VP16_TRANSGENIC_HDAC7_KO_DOUBLE_POSITIVE_THYMOCYTE_UP"]], ranks)

fgseaRes_sig <- subset(fgseaRes, pval < 0.05)
fgseaRes_sig <- fgseaRes_sig[order(fgseaRes_sig$NES,decreasing = TRUE),]
fgseaRes_sig_csv <- as.data.frame(fgseaRes_sig)
# Flatten list columns (e.g., leadingEdge)
fgseaRes_sig_csv$leadingEdge <- sapply(fgseaRes_sig_csv$leadingEdge, paste, collapse = ", ")

# Write to CSV
write.csv(fgseaRes_sig_csv, file = "fgseaRes_CD4_C7gmt_sig.csv", row.names = FALSE)

write.csv(as.data.frame(fgseaRes_sig), file = "fgseaRes_CD4_C7gmt_sig.csv")

topUp <- fgseaRes %>% 
    filter(ES > 0) %>% 
    top_n(10, wt=-padj)
topDown <- fgseaRes %>% 
    filter(ES < 0) %>% 
    top_n(10, wt=-padj)
topPathways <- bind_rows(topUp, topDown) %>% 
    arrange(-ES)
plotGseaTable(C7_Pathways[topPathways$pathway], 
              ranks, 
              fgseaRes, 
              gseaParam = 0.5)

# Visualize GSEA results
plotEnrichment(c7_list[["GSE9988_ANTI_TREM1_VS_CTRL_TREATED_MONOCYTES_UP"]], gene_list) + ggtitle("Enrichment plot: GSE9988_ANTI_TREM1_VS_CTRL_TREATED_MONOCYTES_UP")  # Replace "GSE12345" with an actual gene set ID

# Summarize results
topPathwaysUp <- fgsea_results[ES > 0][order(pval)][1:10, ]
topPathwaysDown <- fgsea_results[ES < 0][order(pval)][1:10, ]
topPathways <- rbind(topPathwaysUp, topPathwaysDown)

# Plot top pathways
library(ggplot2)
ggplot(topPathways, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x = "Pathway", y = "Normalized Enrichment Score", title = "Top GSEA Pathways") +
  theme_minimal()

# Save GSEA results
write.table(fgsea_results, file = "GSEA_C7_fgsea_results.txt", sep = "\t", row.names = FALSE, quote = FALSE)
```

```{r}

################################ FOOT PRINTING
################################ 

library(ATACseqQC)
library(GenomicRanges)
library(Rsamtools)
library(BSgenome.Hsapiens.UCSC.hg38)
library(motifmatchr)
library(chromVAR)
library(SummarizedExperiment)
library(MotifDb)
library(GenomicAlignments)
bamfiles <- list.files(path = "D:\\Kfir\\ATAC-Seq\\R_DiffBind_Analysis\\Bwa-mem_Macs2_Output", pattern = "*.bam$", full.names = TRUE)

# Convert consensus peaks to GRanges object
gr_peaks <- CD4_Counts_norm_contrast_analyzed

# Calculate the Tn5 insertion sites for each BAM file
bamTn5_list <- lapply(bamfiles, function(bamfile) {
  alignments <- readGAlignments(bamfile)
  shifted_alignments <- shiftGAlignmentsList(alignments, genome = BSgenome.Hsapiens.UCSC.hg38, mapqFilter = 30)
  shifted_alignments
})

# Combine the insertion sites into a single object
combined_bamTn5 <- Reduce(c, bamTn5_list)

# Create a combined Footprint Score
combined_footprints <- footPrint(combined_bamTn5, peaks = gr_peaks, genome = BSgenome.Hsapiens.UCSC.hg38)

# Visualize Combined Footprints
plotFootprints(combined_footprints)

IRF2 <- query(MotifDb, c("IRF2"))
IRF2 <- as.list(IRF2)
print(IRF2[[1]], digits=2)


```


```{r}
#Footprinting another guide
library(ATACseqQC)
library(MotifDb)

library(BSgenome.Hsapiens.UCSC.hg38)
genome <- Hsapiens

shifted.bamFile="D:/Kfir/ATAC-Seq/R_DiffBind_Analysis/Bwa-mem_Macs2_Output/G08CD4_S145927_NoMT_subset.blacklist-filtered.shifted.sorted.bam"

library(Rsamtools)
bamHeader <- scanBamHeader(shifted.bamFile)
bamSeqnames <- names(bamHeader[[1]]$targets)
print(bamSeqnames)

PITX3 <- query(MotifDb, c("CTCF"))
PITX3 <- as.list(PITX3)
print(PITX3[[1]], digits=2)

pdf("PITX3_footprnt.pdf")
PITX3 <- factorFootprints(c(shifted.bamFile), pfm=PITX3[[1]],
                         genome=genome,
                         min.score="90%", seqlev=bamSeqnames,
                         upstream=100, downstream=100)
dev.off()

```

```{r}
CD4_DB_DF <- as.data.frame(CD4_Counts_norm_contrast_analyzed.DB)

CD4_DB_DF_merged <- add_gene_names(CD4_DB_DF, CD4_Counts_norm_contrast_analyzed)

write.csv(CD4_DB_DF_merged,"./CD4_Counts_norm_contrast_analyzed_DBS_EDGER_TMM_Names.csv")


```


```{r}
# Assuming your GRanges object is named CD4_Counts_norm_contrast_analyzed.DB

# Create Gain and Loss GRanges objects based on Fold change
gain <- CD4_Counts_norm_contrast_analyzed.DB[CD4_Counts_norm_contrast_analyzed.DB$Fold > 0]
loss <- CD4_Counts_norm_contrast_analyzed.DB[CD4_Counts_norm_contrast_analyzed.DB$Fold < 0]

# Combine the Gain and Loss GRanges objects into a GRangesList
granges_list <- GRangesList(Gain = gain, Loss = loss)

# Check the result
print(granges_list)



profiles <- dba.plotProfile(CD4_Counts_norm_contrast_analyzed,sites = granges_list)#sites = CD4_Counts_norm_contrast_analyzed.DB)
png('./CD4_DBA_PlotProfile.png', units="in", width=7, height=7, res=800)
dba.plotProfile(profiles)
dev.off()

profiles2 <- dba.plotProfile(CD4_Counts_norm_contrast_analyzed, sites=granges_list, contrast=1, 
                bHeatmap=TRUE, bLines=TRUE, 
                labels=c("Good", "Poor"))
png('./CD4_DBA_PlotProfile2.png', units="in", width=7, height=7, res=800)
dba.plotProfile(profiles)
dev.off()
```

```{r}
CD4_overlap_rates <- dba.overlap(CD4, mode=DBA_OLAP_RATE)
plot(CD4_overlap_rates, xlab="Overlapping samples", ylab="Overlapping peaks", type="b")


```


```{r}
############################### CD8 ##################
#CD8_Good_Poor_Results_List <- ATAC_Seq_DBS_Analysis("Sample_Sheet_CD8.csv", "CD8", "good", "poor", usePval = T, pval_th = 0.05, outputdir = "./CD8_GoodvsPoor_Results/")
CD8_Good_Poor_Results_List <- ATAC_Seq_DBS_Analysis("Sample_Sheet_CD8.csv", "CD8", "good", "poor", usePval = T, pval_th = 0.05, outputdir = "./CD8_GoodvsPoor_Results/", dba_counts_RDS_File = "./CD8_GoodvsPoor_Results/CD8_Counts.RDS")

```


```{r}


CD8_Counts_norm_contrast_analyzed <- CD8_Good_Poor_Results_List[[1]]$dba_Counts_norm_contrast_analyzed
CD8_DBS_sig <- CD8_Good_Poor_Results_List[[1]]$DBS_sig_report
CD8_DBS_All <- CD8_Good_Poor_Results_List[[1]]$DBS_All_report

pvals <- dba.plotBox(CD8_Counts_norm_contrast_analyzed,contrast = 1,bUsePval = T ,method = DBA_EDGER)
venn_Gain_Loss<-dba.plotVenn(CD8_Counts_norm_contrast_analyzed, contrast = 1, method = DBA_EDGER,bUsePval =TRUE, bDB=TRUE,bGain=TRUE, bLoss=TRUE, bAll=FALSE, th=0.05,label1 = "Good", label2 = "Poor")

#Extracting Human TF genes
#Downloaded from AnimalTFDB4
tf_data <- read.table("Homo_sapiens_TF_AnimalTFDB4.txt", header = TRUE, sep = "\t")
CD8_DBS_All_TFs <- subset(CD8_DBS_All, CD8_DBS_All$SYMBOL %in% tf_data$Symbol)
write.csv(CD8_DBS_All_TFs[order(CD8_DBS_All_TFs$p.value,decreasing = FALSE),], "./CD8_GoodvsPoor_Results/CD8_Experiment_TF_Genes_ALL_Counts_norm_contrast_analyzed_DBSedgeRGLM_TMM.csv")

```


```{r}
##################################### CD8 GSEA #########################################

CD8_DBS_All_Volcano_Sig <- subset(CD8_DBS_All, Significance == "Significant")
# Remove duplicates by filtering rows with greatest absolute Foldchange for each gene
CD8_DBS_All_Volcano_Sig_NoDups <- CD8_DBS_All_Volcano_Sig %>%
  group_by(SYMBOL) %>%
  slice(which.max(abs(Fold))) %>%
  ungroup()

df_ForGSEA <- subset(CD8_DBS_All, select = c(Fold,p.value,SYMBOL))


C7_Pathways <- gmtPathways("c7.all.v2023.2.Hs.symbols.gmt")

# Prepare the data
ranks <- df_ForGSEA$Fold
names(ranks) <- df_ForGSEA$SYMBOL
head(ranks)
gene_list <- sort(ranks, decreasing = TRUE)
barplot(gene_list)

set.seed(123)
# Perform GSEA
CD8_fgseaRes <- fgsea(pathways = C7_Pathways,
                       stats = gene_list,
                       minSize = 15,
                       maxSize = 1000, eps = 0)#nperm = 10000)

# Create a subset where the pathway includes the substring "CD8"
CD8_fgseaRes_CD8 <- subset(CD8_fgseaRes, grepl("CD8", pathway))
CD8_fgseaRes_CD8 <- CD8_fgseaRes_CD8[order(CD8_fgseaRes_CD8$NES,decreasing = TRUE),]
CD8_fgseaRes_CD8_csv <- CD8_fgseaRes_CD8
CD8_fgseaRes_CD8_csv$leadingEdge <- sapply(CD8_fgseaRes_CD8_csv$leadingEdge, paste, collapse = ", ")

# Optionally, save these subsets to CSV files
write.csv(CD8_fgseaRes_CD8_csv, file = "./CD8_GoodvsPoor_Results/GSEA/CD8_Experiment_fgseaRes_CD8_C7gmt_sig.csv", row.names = FALSE)
  

```


```{r}

#All top pathways up and down regulated
CD8_fgseaRes$Enrichment = ifelse(CD8_fgseaRes$NES > 0, "Up-regulated", "Down-regulated")

topPathwaysUp <- na.omit(CD8_fgseaRes[NES > 0][order(pval)][1:10,])
topPathwaysDown <- na.omit(CD8_fgseaRes[NES < 0][order(pval)][1:10,])
topPathways <- rbind(topPathwaysUp, topPathwaysDown)

# Plot top pathways
#library(ggplot2)
#(topPathways, aes(reorder(pathway, NES), NES)) +
#  geom_col(aes(fill = NES > 0)) +
#  coord_flip() +
#  labs(x = "Pathway", y = "Normalized Enrichment Score", title = "Top GSEA Pathways") +
#  theme_minimal()


colos = setNames(c("firebrick2", "dodgerblue2"),
           c("Up-regulated", "Down-regulated"))
header =  paste0("Top ",nrow(topPathways)," Enriched Pathways In CD8 Experiment (Total pathways: Up=", nrow(topPathwaysUp),", Down=",  nrow(topPathwaysDown), ")")


g1 = ggplot(topPathways, aes(reorder(pathway, NES), NES)) +
geom_point( aes(fill = Enrichment, size = size), shape=21) +
scale_fill_manual(values = colos ) +
scale_size_continuous(range = c(2,10)) +
geom_hline(yintercept = 0) +
coord_flip() +
labs(x="Pathway", y="Normalized Enrichment Score",
 title=header)

png("./CD8_GoodvsPoor_Results/GSEA/CD8_Experiment_topPathWays_GSEA_Plot.png", units="in", width=18, height=18, res=800)
g1
dev.off()
pdf("./CD8_GoodvsPoor_Results/GSEA/CD8_Experiment_topPathWays_GSEA_Plot.pdf", width=18, height=18)
g1
dev.off()
```




```{r}
#CD8 related pathways
#Removing not important pathways:
CD8_fgseaRes_CD8_filt_NonImportant <- CD8_fgseaRes_CD8[-which(CD8_fgseaRes_CD8$pathway %in% CD8_fgseaRes_CD8[c(3,7,9,14,15,16,20,22),]$pathway),]

CD8_fgseaRes_CD8_top30 = CD8_fgseaRes_CD8_filt_NonImportant[1:30,]
CD8_fgseaRes_CD8_top30$Enrichment = ifelse(CD8_fgseaRes_CD8_top30$NES > 0, "Up-regulated", "Down-regulated")

CD8_fgseaRes_CD8_top30 = rbind(head(CD8_fgseaRes_CD8_top30, n = nrow(subset(CD8_fgseaRes_CD8_top30, NES>0))),
                tail(CD8_fgseaRes_CD8_top30, n = nrow(subset(CD8_fgseaRes_CD8_top30, NES<0)) ))
total_up = sum(CD8_fgseaRes_CD8_top30$Enrichment == "Up-regulated")
total_down = sum(CD8_fgseaRes_CD8_top30$Enrichment == "Down-regulated")
header = paste0("Top ",nrow(CD8_fgseaRes_CD8_top30)," Enriched CD8 related Pathways In CD8 Experiment (Total pathways: Up=", total_up,", Down=",  total_down, ")")

colos = setNames(c("firebrick2", "dodgerblue2"),
           c("Up-regulated", "Down-regulated"))

png("./CD8_GoodvsPoor_Results/GSEA/CD8_Experiment_CD8Pathways_top30_GSEA_Plot.png", units="in", width=15, height=15, res=800)
ggplot(CD8_fgseaRes_CD8_top30, aes(reorder(pathway, NES), NES)) +
geom_point( aes(fill = Enrichment, size = size), shape=21) +
scale_fill_manual(values = colos ) +
scale_size_continuous(range = c(2,10)) +
geom_hline(yintercept = 0) +
coord_flip() +
labs(x="Pathway", y="Normalized Enrichment Score",
 title=header)
dev.off()

pdf("./CD8_GoodvsPoor_Results/GSEA/CD8_Experiment_CD8Pathways_top30_GSEA_Plot.pdf", width=15, height=15)
ggplot(CD8_fgseaRes_CD8_top30, aes(reorder(pathway, NES), NES)) +
geom_point( aes(fill = Enrichment, size = size), shape=21) +
scale_fill_manual(values = colos ) +
scale_size_continuous(range = c(2,10)) +
geom_hline(yintercept = 0) +
coord_flip() +
labs(x="Pathway", y="Normalized Enrichment Score",
 title=header)
dev.off()

```














































































```{r}
################################## CD8 ##########################################

# Set the working directory
setwd("D:\\Kfir\\ATAC-Seq\\Bwa-mem_Macs2_Output\\peak_files")

# Load the sample sheet
CD8_samples <- read.csv("Sample_Sheet_CD8.csv")

# Initialize the dba object with the sample sheet
CD8 <- dba(sampleSheet=CD8_samples)
CD8_Counts <- dba.count(CD8, bParallel = TRUE, score=DBA_SCORE_READS)

png('./CD8_Counts_heatmap.png', units="in", width=7, height=7, res=800)
dba.plotHeatmap(CD8_Counts, main = "Correlation Plot for CD8 Samples")
#plot(CD8_Counts, main="Correlation Plot for CD8 Samples")
dev.off()

png('./CD8_counts_PCA.png', units="in", width=7, height=7, res=800)
dba.plotPCA(CD8_Counts,DBA_CONDITION, label=c(DBA_ID))
dev.off()
```

```{r}
CD8_overlap_rates <- dba.overlap(CD8, mode=DBA_OLAP_RATE)
plot(CD8_overlap_rates, xlab="Overlapping samples", ylab="Overlapping peaks", type="b")

```

























































































































































```{r}

######################            ################
######################DGE Analysis################
######################     CD4       ################
library(ChIPQC)
library(soGGi)
peaks <- dir("D:/Kfir/ATAC-Seq/Bwa-mem_Macs2_Output/peak_files", pattern = "CD4.*\\.narrowPeak$", 
    full.names = TRUE)
myPeaks <- lapply(peaks, ChIPQC:::GetGRanges, simple = TRUE)

names(myPeaks) <- c("good","good","poor","good","poor","poor","good","good","poor")
Group <- factor( c("good","good","poor","good","poor","poor","good","good","poor"))
consensusToCount <- soGGi:::runConsensusRegions(GRangesList(myPeaks))




```






```{r}
########################### Heatmap #######################################
# Step 1: Extract the data matrix from the dba object
data_matrix <- dba.plotHeatmap(CD4_Counts_norm_contrast_analyzed, contrast=1, bUsePval=T, th = 1,
                               method=DBA_EDGER, correlations=FALSE, scale="row",)
data_matrix <- as.data.frame(data_matrix)

merged_data <- add_gene_names(data_matrix, CD4_Counts_norm_contrast_analyzed)

merged_data <- subset(merged_data, select = -c(seqnames, start,end,width,strand))
merged_data <- merged_data %>% select("SYMBOL", everything())
#rename(merged_data,c("SYMBOL" = "Gene_name"))

averaged_data <- merged_data %>%
  group_by(SYMBOL) %>%
  summarise(across(everything(), mean, na.rm = TRUE))

averaged_data <- averaged_data %>% remove_rownames %>% column_to_rownames(var="SYMBOL")
# Step 3: Plot the heatmap using ComplexHeatmap
heatmap_colors <- rev(colorRampPalette(brewer.pal(10, "RdBu"))(256))

heatmap_annotation_df <- data.frame(
  Condition = factor(c("Poor","Poor", "Poor", "Poor", "Good", "Good", "Good","Good","Good")),
  row.names = colnames(averaged_data)  # Assuming first column is SYMBOL and rest are samples
)

heatmap_annotation_colors <- list(
  Condition = c(
    "Poor" = "red",    # Replace with your desired color
    "Good" = "blue"    # Replace with your desired color
  )
)


pheatmap_CD4 <- pheatmap::pheatmap(as.matrix(averaged_data %>% `rownames<-`( NULL )),
                                   color = heatmap_colors,
                                   show_row_names = F,
                                   cluster_rows = T,
                                   treeheight_row = 0,
                                   show_column_names = F,
                                   cluster_cols = T,
                                   scale="row",
                                   row_names_gp = gpar(fontsize = 0.5),
                                   annotation_col = heatmap_annotation_df,
                                   annotation_colors = heatmap_annotation_colors) 
pheatmap_CD4

heatmap_toSave_CD4 <- round(pheatmap.scale(averaged_data),2)
heatmap_toSave_CD4 <- heatmap_toSave_CD4[pheatmap_CD4$tree_row$order, pheatmap_CD4$tree_col$order]
write.csv(heatmap_toSave_CD4, "pheatmap_CD4.csv")

```





```{r}
#hmap <- colorRampPalette(c("red", "black", "green"))(n = 13)
#library(RColorBrewer)
readscores <- dba.plotHeatmap(CD4_Counts_norm_contrast_analyzed, contrast=1, bUsePval = T, method = DBA_EDGER, correlations=FALSE,scale="row", colScheme = colorRampPalette(brewer.pal(10,"RdBu"))(256))
write.csv(as.data.frame(readscores),"./CD4_Counts_norm_contrast_analyzed_readscoresHeatmap.csv")

```